datafusion/physical-expr/src/utils/guarantee.rs

```diff
@@ -0,0 +1,699 @@
+                .and(col("b").not_eq(lit(2)))
+                .and(col("b").lt(lit(3))),
+            vec![not_in_guarantee("b", [1, 2])],
+        );
+    }
+
+    #[test]
+    fn test_conjunction_multi_column() {
+        // a = "foo" AND b = 1
+        test_analyze(
+            col("a").eq(lit("foo")).and(col("b").eq(lit(1))),
+            vec![
+                // should find both column guarantees
+                in_guarantee("a", ["foo"]),
+                in_guarantee("b", [1]),
+            ],
+        );
+        // a != "foo" AND b != 1
+        test_analyze(
+            col("a").not_eq(lit("foo")).and(col("b").not_eq(lit(1))),
+            // should find both column guarantees
+            vec![not_in_guarantee("a", ["foo"]), not_in_guarantee("b", [1])],
+        );
+        // a = "foo" AND a = "bar"
+        test_analyze(
+            col("a").eq(lit("foo")).and(col("a").eq(lit("bar"))),
+            // this predicate is impossible ( can't be both foo and bar),
+            vec![],
+        );
+        // a = "foo" AND b != "bar"
+        test_analyze(
+            col("a").eq(lit("foo")).and(col("a").not_eq(lit("bar"))),
+            vec![in_guarantee("a", ["foo"]), not_in_guarantee("a", ["bar"])],
+        );
+        // a != "foo" AND a != "bar"
+        test_analyze(
+            col("a").not_eq(lit("foo")).and(col("a").not_eq(lit("bar"))),
+            // know it isn't "foo" or "bar"
+            vec![not_in_guarantee("a", ["foo", "bar"])],
+        );
+        // a != "foo" AND a != "bar" and a != "baz"
+        test_analyze(
+            col("a")
+                .not_eq(lit("foo"))
+                .and(col("a").not_eq(lit("bar")))
+                .and(col("a").not_eq(lit("baz"))),
+            // know it isn't "foo" or "bar" or "baz"
+            vec![not_in_guarantee("a", ["foo", "bar", "baz"])],
+        );
+        // a = "foo" AND a = "foo"
+        let expr = col("a").eq(lit("foo"));
+        test_analyze(expr.clone().and(expr), vec![in_guarantee("a", ["foo"])]);
+        // b > 5 AND b = 10 (should get an b = 10 guarantee)
+        test_analyze(
+            col("b").gt(lit(5)).and(col("b").eq(lit(10))),
+            vec![in_guarantee("b", [10])],
+        );
+
+        // a != "foo" and (a != "bar" OR a != "baz")
+        test_analyze(
+            col("a")
+                .not_eq(lit("foo"))
+                .and(col("a").not_eq(lit("bar")).or(col("a").not_eq(lit("baz")))),
+            // a is not foo (we can't represent other knowledge about a)
+            vec![not_in_guarantee("a", ["foo"])],
+        );
+    }
+
+    #[test]
+    fn test_conjunction_and_disjunction_single_column() {
+        // b != 1 AND (b > 2)
+        test_analyze(
+            col("b")
+                .not_eq(lit(1))
+                .and(col("b").gt(lit(2)).or(col("b").eq(lit(3)))),
+            vec![
+                // for the expression to be true, b can not be one
+                not_in_guarantee("b", [1]),
```

Comment:
I have to admit that I need to think hard how Bloom filter works to review these. I stared at this for a while and this is what I think. I may misunderstand it though

If the test is `b !=1 and b > 2`, what is the right result here? I think it should be an empty vector, right?
My reasoning is if `b = 0`, then the `not_in_guarantee("b", [1])` is true but the expression `b !=1 and b > 2` is false.

I think it is the same for `b != 1 AND (b > 2 OR b = 3)`

Reply:
The idea for using this for bloom filters is to try and prove that this expression is not true:

```
b != 1 AND (b > 2)
```

In order for the expression to be true, *both* of these have to hold
1.  `b != 1
2. b > 2

Thus, if we can prove either does not hold, we can prove the expression is not true

So therefore, to prove the expression can't possibly be true, we need to prove that b **ONLY** takes the value of `1`.

But as you point out this is not what the documentation says at the moment

```
/// 2. The column must NOT be one of a particular set of values for the predicate to
/// be `true`. If the column takes on any of the values, the predicate can not
/// evaluate to `true`. For example,
/// `(a != 1)`, `(a != 1 AND a != 2)` or `a NOT IN (1, 2, 3)`
```

It should probably say 
```
/// 2. The column must NOT be one of a particular set of values for the predicate to
/// be `true`. If the column can ONLY take one of these values, the predicate can not
/// evaluate to `true`. 
```


🤔  I need to think about this more deeply

Reply:
Here is my understanding: as guarantee is allowed to be false positive, it's fine to drop arbitrary parts of `AND` in a expr -- we only need to make sure the `false` condition is reliable.

Reply:
I tried to clarify more in f65a1ddd9

For the `not_in_garantee`, to prove the expression is false, we have to show that the column can only be one of the set of values.